### ISSUE TITLE
VSPHERE Post Processor to Upload in Parallel. Question/Feature Request

### ISSUE THREAD

**Issue by gamethis** [hashicorp/packer#5847]
Is it on the ROAD MAP to allow the VSPHERE Post Processor to upload templates in parallel.

Feature.
VSPERE Post Processor uploads in parallel.
Durring post processing I upload to about 14 different VSPHERE enviornments.
This takes a huge amount of time to complete.
Would be nice to have a flag for parallel upload that would cause the uploads to vshpere to happen in parallel instead of one at a time.

**Comment by mwhooker** [comment]
Can you post your packer.json so I have a better idea of what you mean? post-processors I *believe* should run in parallel if they are not pipelined. i.e.

```
{
...
"post-processors": [
 // upload-1
 // upload-2
 // etc
]
}
```

**Comment by gamethis** [comment]
{
    "variables": {
        "win_username": null,
        "win_password": null,
        "vsphere_vm_name": null,
        "path": null,
        "vsphere_password": null,
        "vsphere_username": null
    },
    "builders": [{
        "name": "{{user `vsphere_vm_name`}}",
        "type": "vmware-vmx",
        "source_path": "output/{{user `vsphere_vm_name`}}/packer-{{user `vsphere_vm_name`}}.vmx",
        "output_directory": "output/upload/{{user `vsphere_vm_name`}}",
        "communicator": "ssh",
        "ssh_username": "{{user `win_username`}}",
        "ssh_password": "{{user `win_password`}}",
        "ssh_timeout": "240m",
        "ssh_port": "22",
        "headless": "TRUE",
        "floppy_dirs": [
            "{{user `path` }}/floppy/*"
        ],
        "shutdown_timeout": "120m",
        "shutdown_command": "shutdown /s /t 5 /f /d p:4:1 /c \"Packer Shutdown\"",
        "vnc_port_min": 5900,
        "vnc_port_max": 5980,
        "vnc_bind_address": "0.0.0.0",
        "vmx_data": {
            "memsize": "4096",
            "numvcpus": "2",
            "virtualHW.version": "11",
            "scsi0.virtualDev": "pvscsi",
            "mem.hotadd": "TRUE",
            "vcpu.hotadd": "FALSE"
        }
    }],
    "post-processors": [{
            "type": "vsphere",
            "cluster": "MYCLUSTER",
            "datacenter": "MYDATACENTER",
            "datastore": "MYCLUSTER",
            "host": "vshpereserver.somedomain.com",
            "resource_pool": "mypool",
            "vm_folder": "ServerOnDemand/Templates",
            "username": "{{user `vsphere_username`}}",
            "password": "{{user `vsphere_password`}}",
            "vm_network": "vmnetworkinfo",
            "vm_name": "{{user `vsphere_vm_name`}}-MYCLUSTER-MYDATACENTER-dev",
            "disk_mode": "thin",
            "insecure": true
        },
        {
            "type": "vsphere",
            "cluster": "NF04_Farm/MYCLUSTER",
            "datacenter": "MYDATACENTER",
            "datastore": "MYCLUSTER_SOD",
            "host": "vshpereserver.somedeomain.com",
            "resource_pool": "mypool",
            "vm_folder": "ServerOnDemand/Templates",
            "username": "{{user `vsphere_username`}}",
            "password": "{{user `vsphere_password`}}",
            "vm_network": "vmnetworkinfo",
            "vm_name": "{{user `vsphere_vm_name`}}-MYCLUSTER-MYDATACENTER-dev",
            "disk_mode": "thin",
            "insecure": true
        },
        {
            "type": "vsphere",
            "cluster": "NF04_Farm/MYCLUSTER",
            "datacenter": "MYDATACENTER",
            "datastore": "MYCLUSTER_AP",
            "host": "vshpereserver.somedomain.com",
            "resource_pool": "mypool",
            "vm_folder": "ServerOnDemand/Templates",
            "username": "{{user `vsphere_username`}}",
            "password": "{{user `vsphere_password`}}",
            "vm_network": "vmneworkinfo",
            "vm_name": "{{user `vsphere_vm_name`}}-MYCLUSTER-MYDATACENTER-np",
            "disk_mode": "thin",
            "insecure": true
        },
        {
            "type": "vsphere",
            "cluster": "MYCLUSTER",
            "datacenter": "MYDATACENTER",
            "datastore": "MYCLUSTER",
            "host": "vshpereserver.somedomain.com",
            "resource_pool": "mypool",
            "vm_folder": "ServerOnDemand/Templates",
            "username": "{{user `vsphere_username`}}",
            "password": "{{user `vsphere_password`}}",
            "vm_network": "vmneworkinfo",
            "vm_name": "{{user `vsphere_vm_name`}}-MYCLUSTER-MYDATACENTER-np",
            "disk_mode": "thin",
            "insecure": true
        },
        {
            "type": "vsphere",
            "cluster": "MYCLUSTER",
            "datacenter": "MYDATACENTER",
            "datastore": "MYCLUSTER",
            "host": "vshpereserver.somedomain.com",
            "resource_pool": "mypool",
            "vm_folder": "ServerOnDemand/Templates",
            "username": "{{user `vsphere_username`}}",
            "password": "{{user `vsphere_password`}}",
            "vm_network": "vmneworkinfo",
            "vm_name": "{{user `vsphere_vm_name`}}-MYCLUSTER-MYDATACENTER-zone_a",
            "disk_mode": "thin",
            "insecure": true
        },
        {
            "type": "vsphere",
            "cluster": "MYCLUSTER",
            "datacenter": "MYDATACENTER",
            "datastore": "MYCLUSTER",
            "host": "vshpereserver.somedomain.com",
            "resource_pool": "mypool",
            "vm_folder": "ServerOnDemand/Templates",
            "username": "{{user `vsphere_username`}}",
            "password": "{{user `vsphere_password`}}",
            "vm_network": "vmneworkinfo",
            "vm_name": "{{user `vsphere_vm_name`}}-MYCLUSTER-MYDATACENTER-zone_b",
            "disk_mode": "thin",
            "insecure": true
        },
        {
            "type": "vsphere",
            "cluster": "MYCLUSTER",
            "datacenter": "MYDATACENTER",
            "datastore": "MYCLUSTER_AP",
            "host": "vshpereserver.somedomain.com",
            "resource_pool": "mypool",
            "vm_folder": "ServerOnDemand/Templates",
            "username": "{{user `vsphere_username`}}",
            "password": "{{user `vsphere_password`}}",
            "vm_network": "vmneworkinfo",
            "vm_name": "{{user `vsphere_vm_name`}}-MYCLUSTER-MYDATACENTER-zone_a",
            "disk_mode": "thin",
            "insecure": true
        },
        {
            "type": "vsphere",
            "cluster": "MYCLUSTER",
            "datacenter": "MYDATACENTER",
            "datastore": "MYCLUSTER_AP",
            "host": "vshpereserver.somedomain.com",
            "resource_pool": "mypool",
            "vm_folder": "ServerOnDemand/Templates",
            "username": "{{user `vsphere_username`}}",
            "password": "{{user `vsphere_password`}}",
            "vm_network": "vmneworkinfo",
            "vm_name": "{{user `vsphere_vm_name`}}-MYCLUSTER-MYDATACENTER-zone_b",
            "disk_mode": "thin",
            "insecure": true
        },
        {
            "type": "vsphere",
            "cluster": "MYCLUSTER",
            "datacenter": "MYDATACENTER",
            "datastore": "MYCLUSTER",
            "host": "vshpereserver.somedomain.com",
            "resource_pool": "mypool",
            "vm_folder": "ServerOnDemand/Templates",
            "username": "{{user `vsphere_username`}}",
            "password": "{{user `vsphere_password`}}",
            "vm_network": "vmneworkinfo",
            "vm_name": "{{user `vsphere_vm_name`}}-cluster-MYDATACENTER-zone_a",
            "disk_mode": "thin",
            "insecure": true
        },
        {
            "type": "vsphere",
            "cluster": "MYCLUSTER",
            "datacenter": "MYDATACENTER",
            "datastore": "MYCLUSTER",
            "host": "vshpereserver.somedomain.com",
            "resource_pool": "mypool",
            "vm_folder": "ServerOnDemand/Templates",
            "username": "{{user `vsphere_username`}}",
            "password": "{{user `vsphere_password`}}",
            "vm_network": "vmneworkinfo",
            "vm_name": "{{user `vsphere_vm_name`}}-cluster-MYDATACENTER-zone_b",
            "disk_mode": "thin",
            "insecure": true
        },
        {
            "type": "vsphere",
            "cluster": "MYCLUSTER",
            "datacenter": "MYDATACENTER",
            "datastore": "MYDATASTORE",
            "host": "vshpereserver.somedomain.com",
            "resource_pool": "mypool",
            "vm_folder": "ServerOnDemand/Templates",
            "username": "{{user `vsphere_username`}}",
            "password": "{{user `vsphere_password`}}",
            "vm_network": "vmneworkinfo",
            "vm_name": "{{user `vsphere_vm_name`}}-cluster-MYDATACENTER-zone_a",
            "disk_mode": "thin",
            "insecure": true
        },
        {
            "type": "vsphere",
            "cluster": "MYCLUSTER",
            "datacenter": "MYDATACENTER",
            "datastore": "mydatastore",
            "host": "vshpereserver.somedomain.com",
            "resource_pool": "mypool",
            "vm_folder": "ServerOnDemand/Templates",
            "username": "{{user `vsphere_username`}}",
            "password": "{{user `vsphere_password`}}",
            "vm_network": "vmneworkinfo",
            "vm_name": "{{user `vsphere_vm_name`}}-cluster-MYDATACENTER-zone_b",
            "disk_mode": "thin",
            "insecure": true
        }
    ]
}

**Comment by mwhooker** [comment]
ah, I see. I was wrong, they don't run in parallel.

I've updated the ticket to make your request more general. I don't think there's a way to specifically make the vsphere post-processor run in parallel.
